[docs] Add SRS and SDD documents (#373)
Samsung/TICO: docs/requirements.md

## 1. Introduction

### 1.1 Purpose

The purpose of this document is to define the requirements for developing a Python library that enables the direct conversion of PyTorch models into NPU-compatible format (Circle) without external dependencies. The library will also support modern quantization and optimization techniques, improving model performance and developer productivity.

### 1.2 Scope

This project focuses on:

- Providing a direct conversion mechanism from PyTorch to Circle format.
- Enabling model optimization and quantization for NPU deployment.
- Offering user-friendly APIs for developers and researchers to minimize complexity in the model deployment pipeline.
- Ensuring compatibility with evolving NPU compilers and PyTorch versions.

### 1.3 Stakeholders

- **Developers**: Utilize the library to convert and optimize models for NPU execution.
- **Researchers**: Apply advanced quantization and optimization techniques for experimental models.
- **Product Teams**: Benefit from simplified workflows and faster time-to-market.

### 1.4 Definitions and Acronyms

- **PyTorch**: Open-source machine learning framework.
- **NPU**: Neural Processing Unit.

> - **Circle**: Proprietary intermediate representation (IR) format compatible with NPU compilers.

Reviewer: ```suggestion
- **Circle**: Open-source intermediate representation (IR) format compatible with NPU compilers.
```

- **Optimization**: Process of improving model efficiency, such as kernel fusion or operator reordering.

## 2. Functional Requirements

| ID    | Requirement Name                 | Description |
|-------|----------------------------------|-------------|
| RF-1  | PyTorch Model Conversion         | Convert PyTorch models directly into Circle format, ensuring structural integrity and operator compliance. |
| RF-2  | Optimization Algorithm Support   | Support advanced optimization techniques (e.g., kernel fusion, operator reordering), with options for user-defined configurations. |
| RF-3  | Quantization Integration Support | Consume pre-quantized parameters, fold/remove redundant (de)quant ops, quantize missing biases where required, and insert bridging quantize ops on dtype mismatches to keep the graph legal for Circle/NPU. |
| RF-4  | Verification & Debugging Tools   | Offer debugging mode, display detailed error messages, and validate correctness of outputs compared to original PyTorch models. |
| RF-5  | NPU Compiler Compatibility       | Automatically test converted Circle files against NPU compilers and report compatibility results. |

> ## 3. Non-Functional Requirements

Reviewer: RNF-1 -> Which model?
RNF-2 -> What is the original PyTorch model size? How can it be measured?
RNF-3 -> What does fidelity mean? Maybe we should use PEIR and relax the number 99%.
RNF-4 -> "A wide range of" is ambiguous. It would be better to specify target models (ex: Llama, transformer, and possibly some CNNs?.
RNF-5 -> Better to change compatibility with update of torch version and circle schema. TICO converts torch model to circle. Strictly speaking, NPU compiler is not visible to TICO.

Author: I've rewritten the contents the way it sounds more specific. PTAL:)

| ID     | Requirement Name               | Description |
|--------|--------------------------------|-------------|

> | RNF-1  | Performance: Conversion Speed  | Conversion process should complete within approximately 10 seconds. |

Reviewer: ```suggestion
| RNF-1  | Performance: Conversion Speed  | Conversion process should complete within approximately 10 seconds for Llama 3.2 1B. |
```

Author: Isn't it too specific? The target models are described below RNF-4 and the conversion speed is applied to them.

Reviewer 2: Conversion speed might be related with model size. 
How about just 1B instead of specific model?

Reviewer: Llama model is just an example. As @seockho-kim suggested, 1B scale model (or 3B scale model) would be ok.

>  The target models are described below RNF-4

It would be better to write target models here too for self-containedness.

| RNF-2  | Performance: File Size         | Converted Circle files should be equal to or smaller than the serialized PyTorch model size saved via `torch.save(model.state_dict())`. |
| RNF-3  | Reliability                    | Converted model outputs should remain close to the PyTorch reference outputs. The similarity is measured using **PEIR (Peak Error Interval Ratio)** defined as `max(\|a - b\|) / (max(a) - min(a))`. The PEIR must not exceed **3% (0.03)** for any output tensor. |
| RNF-4  | Maintainability: Operator Support | The library should support operator coverage sufficient for modern Transformer and LLM architectures. |

> | RNF-5  | Maintainability: Schema & Framework Updates | Must remain compatible with updates in PyTorch (latest stable & LTS) and the Circle schema specification. |

Reviewer: ```suggestion
| RNF-5  | Maintainability: Schema & Framework Updates | Must remain compatible with recent PyTorch versions (2.5 ~ latest stable) and the Circle schema specification. |
```
